pennylane_sf/gbs.py

```diff
@@ -124,46 +134,112 @@ def execute(self, queue, observables, parameters=None, **kwargs):
 
     # pylint: disable=pointless-statement,expression-not-assigned
     def apply(self, operation, wires, par):
-        self._params, A, _ = par
+        self._params, A, n_mean = par
+        A *= rescale(A, n_mean)
+        self.Z_inv = self._calculate_z_inv(A)
 
         if len(self._params) != self.num_wires:
             raise ValueError(
                 "The number of variable parameters must be equal to the total number of wires."
             )
 
-        self._WAW = self.calculate_WAW(*par)
-        n_mean_WAW = self.calculate_n_mean(self._WAW)
+        self._WAW = self.calculate_WAW(self._params, A)
+
+        if not self.analytic and self.use_cache:
+            op = GraphEmbed(A, mean_photon_per_mode=n_mean / len(A))
+        else:
+            n_mean_WAW = self.calculate_n_mean(self._WAW)
+            op = GraphEmbed(self._WAW, mean_photon_per_mode=n_mean_WAW / len(A))
 
-        op = GraphEmbed(self._WAW, mean_photon_per_mode=n_mean_WAW / len(A))
         op | [self.q[wires.index(i)] for i in wires]
 
         if not self.analytic:
             MeasureFock() | [self.q[wires.index(i)] for i in wires]
 
+    def reset(self):
+        # There is no engine to reset in this case
```

Comment:
Wait, this assumes that the engine was created by passing the arguments `use_cache=True` and `samples=<some_samples_array>`. What if that was not the case, but for example, the `samples`  attribute was `None` and it gets set later?
```python
dev = qml.device('strawberryfields.gbs', wires=4, cutoff_dim=4,
                samples=None, use_cache=True, analytic=False)

# executing a circuit on dev, samples are stored as per `dev.pre_measure`
dev.reset()  # no reset mechanism, dev.eng still remains with samples, state, etc.
```

Reply:
There are two options in non-analytic + caching mode, i.e.,
1. `samples = None, use_cache=True, analytic=False`
2. `samples = arr, use_cache=True, analytic=False`

In option 1, we don't input user generated samples. Instead, when the qnode is first evaluated on the probabilities, we do sampling (according to the `shots` argument which defaults to 1000). These samples should then be cached and re-used on subsequent evaluations. On the subsequent evaluations, we don't want to call `super().reset()` which gets rid of the samples, so we enter this if-statement and simply return nothing.

For 2, we straight away have user input samples that are added to `self.samples`, and we can enter this loop because again we don't want to run `super().reset()` and lose our samples.

TLDR: we don't want to call `super().reset()` else we lose our cached samples.

I think what is confusing is saying "There is no engine to reset in this case", since actually I think that is not true in case 1 as you point out. There, the engine is initialized but we don't need to worry about resetting it because it won't be used again.

I've changed the comment to:
`# In this case, we do not reset because we want to keep our samples`

Reply:
> TLDR: we don't want to call super().reset() else we lose our cached samples.

Right! Actually, it can also be, that the `GBSDevice` grabs the important parts of `super().reset()` and for example never resets `self.samples`.

> don't need to worry about resetting it because it won't be used again.

Is this for sure? What if the user goes for `dev.use_cache=False` and runs another circuit? Although quite an edge case, maybe it's worth disabling the setting of `use_cache` (e.g. creating a `_use_cache` attribute that is set in `__init__` and making `use_cache` a property that returns `_use_cach`). Technically  `dev._use_cach` could be set by the user, but it would not be supported (and the user would have to tweak the code to produce unwanted behaviour).

Reply:
I like it!

Reply:
Added: 161e5c3516fc50c461273b1fc12bb5fa7d1053a5